moveit_experimental/jog_arm/src/jog_arm/jog_arm_server.cpp

```diff
@@ -0,0 +1,726 @@
+  {
+    ROS_ERROR_STREAM_THROTTLE_NAMED(2, "jog_arm_server", ros::this_node::getName()
+                                                             << " Close to a "
+                                                                "position or velocity limit. Halting.");
+
+    halt(new_jt_traj);
+
+    std_msgs::Bool limit_status;
+    limit_status.data = true;
+    warning_pub_.publish(limit_status);
+  }
+
+  if (jog_arm::g_parameters.gazebo)
+    // Spam several redundant points into the trajectory. The first few may be
+    // skipped if the
+    // time stamp is in the past when it reaches the client. Needed for gazebo
+    // simulation.
+    // Start from 2 because the first point's timestamp is already
+    // 1*jog_arm::g_parameters.publish_period
+    point = new_jt_traj.points[0];
+  for (int i = 2; i < 30; i++)
+  {
+    point.time_from_start = ros::Duration(i * jog_arm::g_parameters.publish_period);
+    new_jt_traj.points.push_back(point);
+  }
+
+  // Share with main to be published
+  pthread_mutex_lock(&jog_arm::g_new_traj_mutex);
+  jog_arm::g_new_traj = new_jt_traj;
+  pthread_mutex_unlock(&jog_arm::g_new_traj_mutex);
+}
+
+// Halt the robot
+void JogCalcs::halt(trajectory_msgs::JointTrajectory& jt_traj)
+{
+  for (std::size_t i = 0; i < jt_state_.velocity.size(); i++)
+  {
+    jt_traj.points[0].positions[i] = orig_jts_.position[i];
+    jt_traj.points[0].velocities[i] = 0.;
+  }
+  // Store all zeros in the velocity filter
+  resetVelocityFilters();
+}
+
+// Reset the data stored in filters so the trajectory won't jump when jogging is
+// resumed.
+void JogCalcs::resetVelocityFilters()
+{
+  for (std::size_t i = 0; i < jt_state_.name.size(); i++)
+    velocity_filters_[i].reset(0);  // Zero velocity
+}
+
+// Update joint velocities
+bool JogCalcs::updateJointVels(sensor_msgs::JointState& output, const Eigen::VectorXd& joint_vels) const
+{
+  for (std::size_t i = 0, size = static_cast<std::size_t>(joint_vels.size()); i < size; ++i)
+  {
+    try
+    {
+      output.velocity[i] = joint_vels(static_cast<long>(i));
+    }
+    catch (std::out_of_range e)
+    {
+      ROS_ERROR_STREAM_NAMED("jog_arm_server", ros::this_node::getName() << " Vector lengths do not match.");
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Parse the incoming joint msg for the joints of our MoveGroup
+void JogCalcs::updateJoints()
+{
+  // Check that the msg contains enough joints
+  if (incoming_jts_.name.size() < jt_state_.name.size())
+  {
+    ROS_WARN_NAMED("JogCalcs", "The joint msg does not contain enough "
+                               "joints.");
+    return;
+  }
+
+  // Store joints in a member variable
+  for (std::size_t m = 0; m < incoming_jts_.name.size(); m++)
+  {
+    for (std::size_t c = 0; c < jt_state_.name.size(); c++)
+    {
+      if (incoming_jts_.name[m] == jt_state_.name[c])
+      {
+        jt_state_.position[c] = incoming_jts_.position[m];
+        goto NEXT_JOINT;
```

Comment:
you should never use goto in C++! please refactor

Reply:
Interesting point of view I hadn't heard before. I've always been told to avoid it 100%.

Second opinion @v4hn ?